### PR TITLE
Add Duration::from_micros

### LINKED ISSUE #44400: Consider adding a from_micros for `std::time::Duration`
During the write of an embedded library I needed to make a pause of 10 microseconds. Looking over the [Nightly `std::time::Duration` documentation page](https://doc.rust-lang.org/nightly/std/time/struct.Duration.html) I understood that I had to use the following construction:

```rust
// Wait 10 microseconds
thread::sleep(Duration::new(0, 10_000));
```

The sad part is that this is less clear to me than `Duration::from_micros(10)`. Currently the `from_millis` and `from_secs` functions are implemented but not `from_micros` which could help write cleaner code IMO.

The fix seems pretty easy, maybe if you judge this useful I can try to write my first pull request to the Rust project!

### PR DESCRIPTION
**Pull request by gagath**
This fixes #44400 that explains why it could be useful for embedded designs timing.

### CODE REVIEW

**Comment by mattico** [comment]
This'll need a feature gate (right?).

You'll need to:

- Create a tracking issue (or just use #44400)
- Add a feature to `src/libsyntax/feature_gate.rs`
- Add an entry to the unstable book `src/doc/unstable-book`
- Add an `#[unstable(feature = "", issue = "")]` annotation to the function.

**Comment by sfackler** [comment]
`src/libsyntax/feature_gate.rs` is only used for language features, not library features. I don't think the unstable book needs to be modified anymore either.

File: src/libstd/time/duration.rs [hunk 1]
```diff
@@ -114,6 +116,27 @@ impl Duration {
...
+    /// let duration = Duration::from_micros(1_000_002);
+    ///
+    /// assert_eq!(1, duration.as_secs());
+    /// assert_eq!(2000, duration.subsec_nanos());
+    /// ```
+    #[inline]
+    pub fn from_micros(micros: u64) -> Duration {
+        let secs = micros / MICROS_PER_SEC;
+        let nanos = ((micros % MICROS_PER_SEC) as u32) * NANOS_PER_MICRO;
+        Duration {
```
Comment (SergioBenitez) [inline src/libstd/time/duration.rs]: This should be inlined for consistency.

**Comment by MicroJoe** [comment]
I have commited some changed according to your feedbacks, please let me know if it needs anything else.

File: src/libstd/time/duration.rs [hunk 2]
```diff
@@ -114,6 +116,25 @@ impl Duration {
         let secs = millis / MILLIS_PER_SEC;
         let nanos = ((millis % MILLIS_PER_SEC) as u32) * NANOS_PER_MILLI;
         Duration { secs: secs, nanos: nanos }
+
```
Comment (kennytm) [inline src/libstd/time/duration.rs]: You missed a `}` here.

**Review by kennytm — CHANGES_REQUESTED** [review]

**Comment by MicroJoe** [comment]
Thanks @kennytm, fixed!

File: src/libstd/time/duration.rs [hunk 3]
```diff
@@ -116,6 +118,26 @@ impl Duration {
...
+    /// let duration = Duration::from_micros(1_000_002);
+    ///
+    /// assert_eq!(1, duration.as_secs());
+    /// assert_eq!(2000, duration.subsec_nanos());
+    /// ```
+    #[unstable(feature = "", issue = "")]
+    #[inline]
+    pub fn from_micros(micros: u64) -> Duration {
+        let secs = micros / MICROS_PER_SEC;
+        let nanos = ((micros % MICROS_PER_SEC) as u32) * NANOS_PER_MICRO;
```
Comment (kennytm) [inline src/libstd/time/duration.rs]: Typo. Also, the `issue` must be a number.

```rust
[00:02:49] error[E0425]: cannot find value `NANOS_PER_MICRO` in this scope
[00:02:49]    --> /checkout/src/libstd/time/duration.rs:137:58
[00:02:49]     |
[00:02:49] 137 |         let nanos = ((micros % MICROS_PER_SEC) as u32) * NANOS_PER_MICRO;
[00:02:49]     |                                                          ^^^^^^^^^^^^^^^ did you mean `NANOS_PER_MICROS`?
[00:02:49] 
[00:02:49] error[E0545]: incorrect 'issue'
[00:02:49]    --> /checkout/src/libstd/time/duration.rs:133:5
[00:02:49]     |
[00:02:49] 133 |     #[unstable(feature = "", issue = "")]
[00:02:49]     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:02:49] 
[00:02:51] error: aborting due to 2 previous errors
[00:02:51] 
[00:02:51] error: Could not compile `std`.
```

File: src/libstd/time/duration.rs [hunk 4]
```diff
@@ -116,6 +118,26 @@ impl Duration {
...
+    ///
+    /// ```
+    /// use std::time::Duration;
+    ///
+    /// let duration = Duration::from_micros(1_000_002);
+    ///
+    /// assert_eq!(1, duration.as_secs());
+    /// assert_eq!(2000, duration.subsec_nanos());
+    /// ```
+    #[unstable(feature = "", issue = "44400")]
```
Comment (sfackler) [inline src/libstd/time/duration.rs]: The feature can be called whatever you want `duration_from_micros` is probably fine.

File: src/libstd/time/duration.rs [hunk 5]
```diff
@@ -116,6 +118,26 @@ impl Duration {
         Duration { secs: secs, nanos: nanos }
     }
 
+    /// Creates a new `Duration` from the specified number of microseconds.
+    ///
+    /// # Examples
+    ///
+    /// ```
```
Comment (sfackler) [inline src/libstd/time/duration.rs]: This test will need to enable that feature to build properly.

**Comment by sfackler** [comment]
Seems like a straightforward/reasonable addition!

@rfcbot fcp merge

**Comment by MicroJoe** [comment]
Thanks for the review @sfackler, I have added a feature name as you suggested and used it in the test.

**Pull request merged**